R/block_cv.R

```diff
@@ -0,0 +1,165 @@
+#'
+#' @return A tibble with classes `spatial_block_cv`, `rset`, `tbl_df`, `tbl`,
+#'   and `data.frame`. The results include a column for the data split objects
+#'   and an identification variable `id`.
+#'
+#' @examples
+#' data(Smithsonian, package = "modeldata")
+#' smithsonian_sf <- sf::st_as_sf(Smithsonian,
+#'                                coords = c("longitude", "latitude"),
+#'                                # Set CRS to WGS84
+#'                                crs = 4326)
+#'
+#' block_cv(smithsonian_sf, v = 3)
+#'
+#' @references
+#'
+#' D. R. Roberts, V. Bahn, S. Ciuti, M. S. Boyce, J. Elith, G. Guillera-Arroita,
+#' S. Hauenstein, J. J. Lahoz-Monfort, B. Schröder, W. Thuiller, D. I. Warton,
+#' B. A. Wintle, F. Hartig, and C. F. Dormann. "Cross-validation strategies for
+#' data with temporal, spatial, hierarchical, or phylogenetic structure," 2016,
+#' Ecography 40(8), pp. 913-929, doi: 10.1111/ecog.02881.
+#'
+#' @export
+block_cv <- function(data, method = "random", v = 10, ...) {
+  method <- rlang::arg_match(method)
+
+  if (!"sf" %in% class(data)) {
+    rlang::abort(
+      c(
+        "`block_cv` currently only supports `sf` objects.",
+        "Try converting `data` to an `sf` object via `sf::st_as_sf()`."
+      )
+    )
+  }
+
+  if (sf::st_crs(data) == sf::NA_crs_) {
+    rlang::abort(
+      c(
+        "`block_cv` requires your data to have an appropriate coordinate reference system (CRS).",
+        "Try setting a CRS using `sf::st_set_crs()`."
+      )
+    )
+  }
+
+  grid_box <- sf::st_bbox(data)
+  if (sf::st_is_longlat(data)) {
+    # cf https://github.com/ropensci/stplanr/pull/467
+    # basically: spherical geometry means sometimes the straight line of the
+    # grid will exclude points within the bounding box
+    #
+    # so here we'll expand our boundary by 0.1% in order to always contain our
+    # points within the grid
+    grid_box[1] <- grid_box[1] - abs(grid_box[1] * 0.001)
+    grid_box[2] <- grid_box[2] - abs(grid_box[2] * 0.001)
+    grid_box[3] <- grid_box[3] + abs(grid_box[3] * 0.001)
+    grid_box[4] <- grid_box[4] + abs(grid_box[4] * 0.001)
+  }
+  grid_blocks <- sf::st_make_grid(grid_box, ...)
+  split_objs <- switch(
+    method,
+    "random" = random_block_cv(data, grid_blocks, v)
+  )
+  v <- split_objs$v[[1]]
+  split_objs$v <- NULL
+
+  ## We remove the holdout indices since it will save space and we can
+  ## derive them later when they are needed.
+  split_objs$splits <- map(split_objs$splits, rm_out)
+
+  ## Save some overall information
+  cv_att <- list(v = v)
+
+  new_rset(
+    splits = split_objs$splits,
+    ids = split_objs[, grepl("^id", names(split_objs))],
+    attrib = cv_att,
+    subclass = c("spatial_block_cv", "rset")
+  )
+
+}
+
+random_block_cv <- function(data, grid_blocks, v) {
+  n <- nrow(data)
+
+  block_contains_points <- purrr::map_lgl(
+    sf::st_intersects(grid_blocks, data),
+    function(x) !identical(x, integer(0))
```

Comment:
This element shows up again later and isn't that straightforward to read, so maybe it's worth putting into a helper function with a speaking name?

Reply:
I _think_ I'm understanding you -- `function(x) !identical(x, integer(0))` is now extracted into `sgbp_is_not_empty()`. Let me know if I misunderstood!